### PR TITLE
Extend FakeMysqlDaemon to respect the  parameters "tables", "excludeTabl...

### PR DESCRIPTION
**Pull request by michael-berlin**
...es" and "includeViews".

Moved existing code out into a new member function SchemaDefinition.FilterTables(...).

Updated Mysqld.GetSchema(...) to get all tables from MySQL first and then filter the tables. Breaking up this work into two phases adds another iteration of the TableDefinitions list and another copy of it.

@aaijazi

### CODE REVIEW

File: go/vt/mysqlctl/schema.go [hunk 1]
```diff
@@ -152,6 +104,10 @@ func (mysqld *Mysqld) GetSchema(dbName string, tables, excludeTables []string, i
 		sd.TableDefinitions = append(sd.TableDefinitions, td)
 	}
 
+	sd, err = sd.FilterTables(tables, excludeTables, includeViews)
+	if err != nil {
+		return nil, err
+	}
 	sd.GenerateSchemaVersion()
```
Comment (aaijazi) [inline go/vt/mysqlctl/schema.go]: Isn't FilterTables already doing GenerateSchemaVersion()? Any reason to do it twice?

File: go/vt/mysqlctl/mysql_daemon.go [hunk 2]
```diff
@@ -101,12 +101,32 @@ func (fmd *FakeMysqlDaemon) SlaveStatus() (*proto.ReplicationStatus, error) {
 	return fmd.CurrentSlaveStatus, nil
 }
 
+// ExcludeTables copies the current schema, excluding tables specified by name.
+func ExcludeTables(sd *proto.SchemaDefinition, excludeTables []string) *proto.SchemaDefinition {
```
Comment (aaijazi) [inline go/vt/mysqlctl/mysql_daemon.go]: This can be removed now...

File: go/vt/mysqlctl/proto/schema_test.go [hunk 3]
```diff
@@ -177,3 +177,115 @@ func TestSchemaDiff(t *testing.T) {
...
+		input         *SchemaDefinition
+		tables        []string
+		excludeTables []string
+		includeViews  bool
+		want          *SchemaDefinition
+	}{
+		{
+			desc: "filter based on tables (whitelist)",
+			input: &SchemaDefinition{
+				DatabaseSchema: "CREATE DATABASE {{.DatabaseName}}",
```
Comment (aaijazi) [inline go/vt/mysqlctl/proto/schema_test.go]: I believe you can drop the DatabaseSchema field for the SchemaDefinitions being used here; they shouldn't be needed for this test.

File: go/vt/mysqlctl/proto/schema_test.go [hunk 4]
```diff
@@ -177,3 +177,115 @@ func TestSchemaDiff(t *testing.T) {
...
+	var testcases = []struct {
+		desc          string
+		input         *SchemaDefinition
+		tables        []string
+		excludeTables []string
+		includeViews  bool
+		want          *SchemaDefinition
+	}{
+		{
+			desc: "filter based on tables (whitelist)",
```
Comment (aaijazi) [inline go/vt/mysqlctl/proto/schema_test.go]: Cool, I like the inclusion of a desc on the tests!

File: go/vt/mysqlctl/proto/schema_test.go [hunk 5]
```diff
@@ -177,3 +177,115 @@ func TestSchemaDiff(t *testing.T) {
...
+					basicTable2,
+				},
+				Version: "6d1d294def9febdb21b35dd19a1dd4c6",
+			},
+		},
+	}
+
+	for _, tc := range testcases {
+		got, err := tc.input.FilterTables(tc.tables, tc.excludeTables, tc.includeViews)
+		if err != nil {
```
Comment (aaijazi) [inline go/vt/mysqlctl/proto/schema_test.go]: These tests look good. You could also supplement them with some tests of error conditions - I think invalid regexes that don't compile might be the only ones.

**Comment by aaijazi** [comment]
LGTM with comments addressed! Add tests for errors is a nice to have, not a blocker.

File: go/vt/mysqlctl/mysql_daemon.go (hunk 2 above)
Comment (michael-berlin) [inline go/vt/mysqlctl/mysql_daemon.go]: DONE

File: go/vt/mysqlctl/proto/schema_test.go (hunk 3 above)
Comment (michael-berlin) [inline go/vt/mysqlctl/proto/schema_test.go]: DONE

File: go/vt/mysqlctl/proto/schema_test.go (hunk 4 above)
Comment (michael-berlin) [inline go/vt/mysqlctl/proto/schema_test.go]: ACK :)

File: go/vt/mysqlctl/proto/schema_test.go (hunk 5 above)
Comment (michael-berlin) [inline go/vt/mysqlctl/proto/schema_test.go]: DONE

Added "wantError" to the test struct and two tests for invalid regex on tables and excludeTables.

File: go/vt/mysqlctl/schema.go (hunk 1 above)
Comment (michael-berlin) [inline go/vt/mysqlctl/schema.go]: FilterTables() only generates the schema version for the new copy of the schema if the original schema had a version. If no Version was set, no version will be generated.

This way FilterTables() and GenerateSchemaVersion() are independent

In this code, "sd" has no version yet and therefore FilterTables() does not set it. Therefore, it's correct to generate the schema version after FilterTables().

**Pull request merged**